packages/backend/src/core/PageCollector.ts

```diff
@@ -0,0 +1,89 @@
+import { BigNumber, utils } from 'ethers'
+
+import {
+  PageRecord,
+  PageRepository,
+} from '../peripherals/database/PageRepository'
+import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
+import { BlockNumber, BlockRange } from '../peripherals/ethereum/types'
+
+const REGISTRY_ABI = new utils.Interface([
+  'event LogMemoryPageFactContinuous(bytes32 factHash, uint256 memoryHash, uint256 prod)',
+])
+
+const REGISTRY_ADDRESS = '0xEfbCcE4659db72eC6897F46783303708cf9ACef8'
+
+const PAGE_ABI = new utils.Interface([
+  'function registerContinuousMemoryPage(uint256 startAddr, uint256[] values, uint256 z, uint256 alpha, uint256 prime)',
+])
+
+/** @internal exported only for tests */
+export const LOG_MEMORY_PAGE_FACT_CONTINUOUS = REGISTRY_ABI.getEventTopic(
+  'LogMemoryPageFactContinuous'
+)
+
+export class PageCollector {
+  constructor(
+    private readonly ethereumClient: EthereumClient,
+    private readonly pageRepository: PageRepository
+  ) {}
+
+  async collect(blockRange: BlockRange): Promise<PageRecord[]> {
+    const memoryPageEvents = await this.getMemoryPageEvents(blockRange)
+
+    const records = await Promise.all(
+      memoryPageEvents.map(
+        async ({ memoryHash, transactionHash }): Promise<PageRecord> => {
+          const tx = await this.ethereumClient.getTransaction(transactionHash)
+
+          const decoded = PAGE_ABI.decodeFunctionData(
```

Comment:
This isn't tested. Instead of having a huge blob of data in tests i suggest using `PAGE_API.encodeFunctionData` and checking that the values are actually what is being decoded as a page.